src/core/solver.py

```
@@ -272,69 +312,288 @@ def add_scoring_objectives(self, strategy: ScoringStrategy) -> None:
+            identity_buckets.keys(), key=lambda k: identity_buckets[k][0]
+        )
+
+        for identity in sorted_bucket_keys:
+            indices = identity_buckets[identity]
+            valid_hinted_g_idxs = sorted(
+                hinted_groups[idx] for idx in indices if idx in hinted_groups
+            )
+
+            if len(indices) > len(valid_hinted_g_idxs):
+                # Anonymize identity to prevent leaking sensitive scores/tags
+                safe_id = hashlib.md5(str(identity).encode()).hexdigest()[:8]
```

Ruff (0.15.12)

[error] Probable use of insecure hash functions in `hashlib`: `md5`

(S324)